### PR TITLE
 [REF] *: dynamicInfo receive dynamicInfo

### PR DESCRIPTION
**Pull request by FrancoisGe**
Before this commit, the use of multiple <field> with the same name in a
view was not well supported.

Why was this?
Some Field components need to know information related to the <field>
such as context, domain, required and readonly. The solution used before
this commit to access this information is to use the getFieldContext,
getFieldDomain, isReadonly, isRequired functions of the model.
Unfortunately, these only take into account the last occurrence of the
<field> because the model is not aware that the same field is present
several times on the view. The information must therefore not come from
the model. For example, it was not possible to have the same field
twice with 2 different domains. It will use the domain of the last
field for both.

Solution:
We will add the object "dynamicInfo" to the fieldInfo passed to the Fields
extractProps function. This object will contain a getter to get the value
of required, readonly, domain and context for the current <field>.
If a Field needs one of its information, it will just have to get it
from extractProps.

Part of Task: 3179751

### CODE REVIEW

File: addons/web/static/src/views/basic_relational_model.js [hunk 1]
```diff
@@ -446,19 +451,16 @@ export class Record extends DataPoint {
...
     }
 
-    getFieldDomain(fieldName) {
-        const domain = this.model.__bm__.localData[this.__bm_handle__].getDomain({
+    getFieldContext(fieldName) {
+        return this.model.__bm__.localData[this.__bm_handle__].getContext({
             fieldName,
             viewType: this.__viewType,
         });
-        return new Domain(JSON.parse(JSON.stringify(domain))); // legacy pyjs inserts weird structures
```
Comment (aab-odoo) [inline addons/web/static/src/views/basic_relational_model.js:461]: also remove isReadonly?

Comment (aab-odoo) [reply addons/web/static/src/views/basic_relational_model.js:461]: and isRequired

File: addons/web/static/src/views/fields/many2many_tags_avatar/many2many_tags_avatar_field.js [hunk 2]
```diff
@@ -139,19 +140,15 @@ export class KanbanMany2ManyTagsAvatarFieldTagsList extends TagsList {
 export class KanbanMany2ManyTagsAvatarField extends Many2ManyTagsAvatarField {
     static template = "web.KanbanMany2ManyTagsAvatarField";
     static components = {
-        ...Many2ManyTagsAvatarField.component,
```
Comment (aab-odoo) [inline addons/web/static/src/views/fields/many2many_tags_avatar/many2many_tags_avatar_field.js:142]: fix for 16.2?

File: addons/web/static/src/views/list/list_renderer.js [hunk 3]
```diff
@@ -796,8 +796,8 @@ export class ListRenderer extends Component {
 
     getCellReadonly(column, record) {
```
Comment (aab-odoo) [inline addons/web/static/src/views/list/list_renderer.js:797]: isCellReadonly would be more appropriate, imho

File: addons/web/static/tests/views/fields/many2one_field_tests.js [hunk 4]
```diff
@@ -3760,7 +3760,7 @@ QUnit.module("Fields", (hooks) => {
         await click(target, ".o_field_widget[name='trululu'] input");
     });
 
-    QUnit.test("many2one in one2many: domain updated by an onchange", async function (assert) {
+    QUnit.skip("many2one in one2many: domain updated by an onchange", async function (assert) {
```
Comment (aab-odoo) [inline addons/web/static/tests/views/fields/many2one_field_tests.js:3763]: ?

File: addons/web/static/tests/views/form/form_view_tests.js [hunk 5]
```diff
@@ -11012,7 +11012,7 @@ QUnit.module("Views", (hooks) => {
         }
     );
 
-    QUnit.test("domain returned by onchange is cleared on discard", async function (assert) {
+    QUnit.skip("domain returned by onchange is cleared on discard", async function (assert) {
```
Comment (aab-odoo) [inline addons/web/static/tests/views/form/form_view_tests.js:11015]: can be removed

File: addons/web/static/tests/views/form/form_view_tests.js [hunk 6]
```diff
@@ -8304,7 +8304,7 @@ QUnit.module("Views", (hooks) => {
         );
     });
 
-    QUnit.test("fields and record contexts are not mixed", async function (assert) {
+    QUnit.skip("fields and record contexts are not mixed", async function (assert) {
```
Comment (aab-odoo) [inline addons/web/static/tests/views/form/form_view_tests.js:8307]: ?

File: addons/web/static/tests/views/list_view_tests.js [hunk 7]
```diff
@@ -2647,7 +2647,11 @@ QUnit.module("Views", (hooks) => {
         assert.containsNone(target, ".o_selected_row");
         await click(target.querySelector(".o_data_row .o_data_cell .o_field_boolean_toggle div"));
         assert.containsOnce(target, ".o_selected_row");
-        await click(target.querySelector(".o_selected_row .o_field_boolean_toggle div"));
+        assert.verifySteps([]);
+
+        // save
+        await clickSave(target);
+        assert.containsNone(target, ".o_selected_row");
```
Comment (aab-odoo) [inline addons/web/static/tests/views/list_view_tests.js:2654]: ?

File: addons/web_editor/static/src/js/backend/html_field.js [hunk 8]
```diff
@@ -683,6 +684,7 @@ export const htmlField = {
             wrapper: options.wrapper,
 
             wysiwygOptions,
+            hasReadonlyModifiers: evaluated.readonly,
```
Comment (aab-odoo) [inline addons/web_editor/static/src/js/backend/html_field.js:687]: would name this readonly like everywhere

**Review by aab-odoo — COMMENTED** [review]
nice :-)

File: addons/web/static/tests/views/list_view_tests.js (hunk 7 above)
Comment (FrancoisGe) [reply addons/web/static/tests/views/list_view_tests.js:2654]: check again

File: addons/web/static/src/views/fields/field.js [hunk 9]
```diff
@@ -121,13 +124,49 @@ export class Field extends Component {
...
+            fieldInfo = {
+                ...fieldInfo,
+                evaluated: {
+                    get context() {
+                        const evalContext = record.getEvalContext
+                            ? record.getEvalContext(false)
+                            : record.evalContext;
+                        const context = omit(
+                            record.context,
+                            ...Object.keys(record.context).filter((key) =>
```
Comment (FrancoisGe) [inline addons/web/static/src/views/fields/field.js:142]: remove ...

File: addons/web/static/src/views/fields/field.js [hunk 10]
```diff
@@ -121,13 +124,49 @@ export class Field extends Component {
...
                 };
             }
+            fieldInfo = {
+                ...fieldInfo,
+                evaluated: {
+                    get context() {
+                        const evalContext = record.getEvalContext
+                            ? record.getEvalContext(false)
+                            : record.evalContext;
+                        const context = omit(
```
Comment (FrancoisGe) [inline addons/web/static/src/views/fields/field.js:140]: do a loop for with delete

File: addons/web/static/src/views/fields/field.js [hunk 11]
```diff
@@ -121,14 +124,47 @@ export class Field extends Component {
...
+                },
+                get required() {
+                    return (
+                        fieldInfo.modifiers &&
+                        fieldInfo.modifiers.required &&
+                        evalDomain(fieldInfo.modifiers.required, record.evalContext)
+                    );
+                },
+            };
+            propsFromNode = this.field.extractProps
```
Comment (aab-odoo) [inline addons/web/static/src/views/fields/field.js:165]: suggestion:
if (this.field.extractProps) {
    const dynamicInfo = {...};
    propsFromNode = this.field.extractProps(fieldInfo, dynamicInfo);
}

**Review by aab-odoo — APPROVED** [review]
robodoo r+

**Pull request closed without merging**